### PR TITLE
Adding syntax tree nodes

### PR DESCRIPTION
**Pull request by GreenSeal**
This patch supports syntax tree nodes architecture.

### CODE REVIEW

File: syntax_tree_nodes/include/types.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+};
+
+class ArrayType : public Type {
+    prim_type_t elt_type_;
+    size_t size;
+};
+
+class FuncType : Type {
+    std::vector<std::unique_ptr<Type>> arg_types_;
```
Comment (FixPlus) [inline syntax_tree_nodes/include/types.h]: We should use non-owning pointer here, because we only reference an already existing type object

Comment (FixPlus) [reply syntax_tree_nodes/include/types.h]: (and in all other places below too)

File: syntax_tree_nodes/include/syntax_node.h [hunk 2]
```diff
@@ -0,0 +1,10 @@
+//
+// Created by denist on 11/14/22.
+//
+
+#include "types.h"
+
+class SyntaxNode {
+    SyntaxNode *parent_;
+    syntax_node_t syntax_node_type_;
```
Comment (FixPlus) [inline syntax_tree_nodes/include/syntax_node.h]: I guess we should provide common interface for children nodes handling as well as parent one in some subclass

File: syntax_tree_nodes/include/statements.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+     * We could use not just vector of Statements, but create another
+     * class Scope, which can also contains, for example, SymbolTable
+     */
+};
+
+class ForLoop : public LoopStmt {
+    std::unique_ptr<Expression> init_expr_;
+    std::unique_ptr<Expression> condition_;
+    std::unique_ptr<Expression> after_loop_expr_;
+    std::vector<std::unique_ptr<Statement>> body_;
```
Comment (FixPlus) [inline syntax_tree_nodes/include/statements.h]: I think we should add separate node for compound statemets and have only one child node for body

File: syntax_tree_nodes/include/declarations.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+class ArrayDecl : virtual public Declaration {
+    ArrayType array_type_
+};
+
+class StructDecl : virtual public Declaration {
+    StructType struct_type_;
+};
+
+class FuncDecl : virtual public Declaration {
+    FuncType func_type_;
```
Comment (FixPlus) [inline syntax_tree_nodes/include/declarations.h]: func_type_ should be a reference to an entry in Type table. (as well as in all cases above)

File: syntax_tree_nodes/include/declarations.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
+//
```
Comment (FixPlus) [inline syntax_tree_nodes/include/declarations.h]: Use some namespace over all class/struct declarations/definitions

**Review by FixPlus — CHANGES_REQUESTED** [review]

File: syntax_tree_nodes/include/declarations.h [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        return var_type_;
+    }
+
+    size_t GetBitLength() const {
+        return bit_length_;
+    }
+
+protected:
+    const VarType *var_type_;
+    size_t bit_length_;
```
Comment (FixPlus) [inline syntax_tree_nodes/include/declarations.h]: I think bit length need to be part of the type

File: syntax_tree_nodes/include/declarations.h [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        return func_type_;
+    }
+
+    const Statement *GetBodyStmtAt(size_t idx) const {
+        return body_.at(idx).get();
+    }
+
+protected:
+    const FuncType *func_type_;
+    std::vector<std::unique_ptr<Statement>> body_;
```
Comment (FixPlus) [inline syntax_tree_nodes/include/declarations.h]: it does not need to be here, we have ChildSyntaxNode that handles children

**Review by FixPlus — CHANGES_REQUESTED** [review]

**Comment by FixPlus** [comment]
After a fix I will merge

File: syntax_tree_nodes/include/types.h (hunk 1 above)
Comment (GreenSeal) [reply syntax_tree_nodes/include/types.h]: Done

File: syntax_tree_nodes/include/syntax_node.h (hunk 2 above)
Comment (GreenSeal) [reply syntax_tree_nodes/include/syntax_node.h]: Done

File: syntax_tree_nodes/include/statements.h (hunk 3 above)
Comment (GreenSeal) [reply syntax_tree_nodes/include/statements.h]: Done

File: syntax_tree_nodes/include/declarations.h (hunk 4 above)
Comment (GreenSeal) [reply syntax_tree_nodes/include/declarations.h]: Done

File: syntax_tree_nodes/include/declarations.h (hunk 6 above)
Comment (GreenSeal) [reply syntax_tree_nodes/include/declarations.h]: Done

File: syntax_tree_nodes/include/declarations.h (hunk 7 above)
Comment (GreenSeal) [reply syntax_tree_nodes/include/declarations.h]: Done

File: syntax_tree_nodes/include/declarations.h (hunk 5 above)
Comment (GreenSeal) [reply syntax_tree_nodes/include/declarations.h]: Done

**Review by FixPlus — APPROVED** [review]
lgtm

**Pull request merged**